src/libs/ActiveClientManager/index.js

```diff
@@ -6,38 +12,61 @@ import * as ActiveClients from '../actions/ActiveClients';
 
 const clientID = Str.guid();
 const maxClients = 20;
-
-let activeClients;
-
-let resolveIsReadyPromise;
-const isReadyPromise = new Promise((resolve) => {
-    resolveIsReadyPromise = resolve;
+let addedSelf = false;
+let activeClients = [];
+let resolveInitedPromise;
+const initedPromise = new Promise((resolve) => {
+    resolveInitedPromise = resolve;
+});
+let resolveSavedSelfPromise;
+const savedSelfPromise = new Promise((resolve) => {
+    resolveSavedSelfPromise = resolve;
 });
 
 /**
+ * Determines when the client is ready. We need to wait both till we saved our ID in onyx AND the init method was called
  * @returns {Promise}
  */
 function isReady() {
-    return isReadyPromise;
+    return Promise.all([initedPromise, savedSelfPromise]);
```

Comment:
It seems to me like the `initedPromise` isn't really doing anything. Is it necessary? It feels like to me that `saveSelfPromise` is the only one that is needed (there is no way it can ever resolve without the lib initialized).

Reply:
We don't want to say that it is ready if we have not been instructed to initialize. We call init after the UI is ready.

Reply:
Hm, so you're saying that `savedSelfPromise` can be resolved before the lib is initialized? But if that's true, then why does initialization matter? Initialization doesn't actually do anything, it just resolves a promise. I think the only thing that matters is that `savedSelfPromise` has been done.

Reply:
It does not do anything in itself, it's indicating that everything is ready. If we did not have this, then it's possible that savedSelfPromise resolves first and starts returning isReady when the UI has not been initialized. 
I tried doing what you suggest and things did not work (in some cases, I think it was hard to reliably reproduce since the order of events is not ensured). What I saw happening is that the saved network request would not fire immediatelly, but instead the next time the sequential network code fired.

Reply:
> it's indicating that everything is ready

Isn't this precisely what `savedSelfPromise` is for though?

> and starts returning isReady when the UI has not been initialized

I don't understand why `ActiveClients` cares if the UI has been initialized or not. Is something else relying on that? This kind of feels like something is being held wrong. Like, `ActiveClients` should not be the lib that is telling other things that the UI is initialized.

Reply:
There's 2 things:
1. We need the list of activeClients to be updated before we start doing things (otherwise 2 tabs could send the same request)
2. We need to wait till the UI is ready before sending the any requests. Not sure exactly why, but this is what we did before, we did not do anything till init was called, even if the client was already saved.

> ActiveClients should not be the lib that is telling other things that the UI is initialized

It is not, it's the other way around, other code is telling ActiveClients that the UI is initialized, same as before.

Reply:
> we already agreed that calling init() does NOTHING

No, we did not. We agreed it was signaling when the UI is ready, so that isReady does not return true till it is.

> 2 does not. I'd like to get to the bottom of that rather than just adding a second promise and assuming that what was there before needs to stay there, especially when it's not fully understood.

Fair. I did try what you said and observed what I said above (queue not firing immediately after load, like it should).
Do you want me to delve more into it and figure out why that was happening and why we added the init method in the first place?

Reply:
One thing I want to clarify: this is broken right now, I have the suspicion is why we are seeing reportID collisions.
So, maybe we can leave as is and investigate all that as a follow up.

Reply:
Yes, could you delve into it more? I'd like to make sure we have a full understanding of this before changing it. When I look at the `init()` method in main:

```
function init() {
    ActiveClients.addClient(clientID)
        .then(resolveIsReadyPromise);
}
```

It's adding the current client, then resolving the ready promise. That tells me that the only thing that matters for the lib to be initialized is that the current client has been added. I don't think there is anything else beyond that.

I'm not sure about the reportID collisions. That's the first I've heard about it, so I don't know why that would imply that activeClients is broken.

Reply:
> I'm not sure about the reportID collisions. That's the first I've heard about it, so I don't know why that would imply that activeClients is broken.

I am 99.9% sure activeClients is broken since we are calling merge on an array, which always replaces the first index of the array. Maybe it is not causing the [reportID collision](https://github.com/Expensify/Expensify/issues/228658), but it could.

> It's adding the current client, then resolving the ready promise. That tells me that the only thing that matters for the lib to be initialized is that the current client has been added. I don't think there is anything else beyond that.

If that's the case, why didn't we save the activeClient directly in this file? Why would we need to tell it to "init"?

Reply:
BTW, I will delve into this tomorrow.
Going to cc @AndrewGable and @marcaaron here, because I see you both worked on this code and might have some context.

Reply:
Looking at past commits, maybe it was because we wanted to [verify the authToken first](https://github.com/Expensify/App/commit/a06d2f2e9b2fbf2bd05a89c1fd23a4af5990d67c#diff-87897d9291cc82037a7d77b5c5196619a383acebf46cf631ae84192a0efa480aR34)? Though not sure if that exists anymore...

Reply:
Trying to follow this conversation but a little lost so just going to write a bunch of stuff about how things work now...

>I am 99.9% sure activeClients is broken since we are calling merge on an array, which always replaces the first index of the array

This is not true for a pure array value like `activeClients`. We concat the value onto the existing array values:

https://github.com/Expensify/react-native-onyx/blob/98260f1deacbcca1a687441164cd45d7d915afb8/lib/Onyx.js#L653-L659

If we look at `main` the way this is intended to work is:

1. User gets the active clients in the initial collection (pruning the head of the list if necessary)
2. We then add the client to the end of the existing list with the `merge()` call
3. We then resolve the "is ready" promise so that anything that needs to check if the current client is the leader can do so with accuracy.

The reasoning behind 3 is if we call `merge()` and wait for it to finish we can also guarantee the `Onyx.connect()` callback has been called again with the updated value for `activeClients`.

As for whether there is some reason we are setting the client in `<Expensify/>` vs. somewhere else. I guess my first question is... where else should it be and what would make that a better place? Does it matter that we are waiting for that component to get constructed? We are going to "wait" until the active client is set at least once. We could probably do this in the `setup.js` code where `Onyx.init()` is called or anywhere early in the app start process. 

Noted that we also set the client again when the app state changes (which presumably happens when you switch to a new tab) and the listener for that is setup in `<Expensify/>` so my guess is that we just thought it was a logical place to stick this.

AFAIK, the only reason we care about the client being the leader is so we can avoid making duplicate requests  and some logic to prevent duplicate browser notifications.

>the only thing that matters for the lib to be initialized is that the current client has been added

Added _and_ the local variable for `activeClients` updated - which we can only guarantee by awaiting the `merge()`.

Reply:
ok, updated and retested this. I indeed removed the initialize promise.